java/src/main/native/src/TableJni.cpp

```diff
@@ -244,6 +244,11 @@ class jni_arrow_output_stream final : public arrow::io::OutputStream {
     if (this->callback == nullptr) {
       throw cudf::jni::jni_exception("global ref");
     }
+
+    this->host_memory_allocator = env->NewGlobalRef(host_memory_allocator);
```

Comment:
Where is this reference destroyed?  NewGlobalRef references must be explicitly disposed by calling DeleteGlobalRef.